base/abstractarray.jl

```diff
@@ -731,19 +730,23 @@ of `IA`.
 
 See also [`checkbounds`](@ref).
 """
-function checkbounds_indices(::Type{Bool}, IA::Tuple, I::Tuple)
+function checkbounds_indices(::Type{Bool}, inds::Tuple, I::Tuple)
     @inline
-    checkindex(Bool, IA[1], I[1])::Bool & checkbounds_indices(Bool, tail(IA), tail(I))
-end
-function checkbounds_indices(::Type{Bool}, ::Tuple{}, I::Tuple)
-    @inline
-    checkindex(Bool, OneTo(1), I[1])::Bool & checkbounds_indices(Bool, (), tail(I))
+    if (N = ndims_cartesian_index(I[1])) === nothing
+        inds₁ = get(inds, 1, OneTo(1))
+        rest = safe_tail(inds)
+    else
+        inds₁, rest = IteratorsMD.split(inds, N)
+    end
+    return checkindex(Bool, inds₁, I[1])::Bool & checkbounds_indices(Bool, rest, tail(I))
 end
-checkbounds_indices(::Type{Bool}, IA::Tuple, ::Tuple{}) = (@inline; all(x->length(x)==1, IA))
+
+checkbounds_indices(::Type{Bool}, inds::Tuple, ::Tuple{}) = (@inline; all(map(isone∘length, inds)))
 checkbounds_indices(::Type{Bool}, ::Tuple{}, ::Tuple{}) = true
 
-throw_boundserror(A, I) = (@noinline; throw(BoundsError(A, I)))
+ndims_cartesian_index(::Any) = nothing # indexes are non-cartesian by default
```

Comment:
Instead of introducing yet another way of getting this information, it'd be good to lean on the existing `index_ndims` functionality.

Reply:
The problem here is we need to check whether the input supports linear indexing.

`checkbounds(zeros(2, 2), CartesianIndex(1))` throws, but `checkbounds(zeros(2, 2), 1)` is OK.
`index_ndims` can't distinguish them. Unfortunately, we'd better not change it's behavior, as it seems to be used widely outside Base.

Reply:
Aha, that makes way more sense to me now. OK, perhaps this should be _supports_linear_indexing, returning a Bool? I don't think we need the counting-dimensions behaviors to be mixed up in this, do we?